dataverse-webapp/src/main/java/edu/harvard/iq/dataverse/users/MergeInAccountService.java

```diff
@@ -0,0 +1,240 @@
+        updateExplicitGroup_AuthenticatedUser(consumedAuthenticatedUser, baseAuthenticatedUser);
+        updateAcceptedConsents(consumedAuthenticatedUser, baseAuthenticatedUser);
+
+        removeConfirmEmailData(consumedAuthenticatedUser);
+        removeOAuth2TokenData(consumedAuthenticatedUser);
+        removeApiToken(consumedAuthenticatedUser);
+        removeUser(consumedAuthenticatedUser);
+    }
+
+    // -------------------- PRIVATE --------------------
+
+    private void updateRoleAssignments(AuthenticatedUser consumedAU, AuthenticatedUser baseAU) {
+        List<RoleAssignment> baseRAList = roleAssigneeService.getAssignmentsFor(baseAU.getIdentifier());
+        List<RoleAssignment> consumedRAList = roleAssigneeService.getAssignmentsFor(consumedAU.getIdentifier());
+
+        for(RoleAssignment consumedRA : consumedRAList) {
+            if(consumedRA.getAssigneeIdentifier().charAt(0) == '@') {
+
+                boolean willDelete = false;
+                for(RoleAssignment baseRA : baseRAList) {
+                    //Matching on the id not the whole DVObject as I'm suspicious of dvobject equality
+                    if (baseRA.getDefinitionPoint().getId().equals(consumedRA.getDefinitionPoint().getId())
+                            && baseRA.getRole().equals(consumedRA.getRole())) {
+                        willDelete = true; //more or less a skip, as we run a delete query afterwards
+                        break;
+                    }
+                }
+                if(!willDelete) {
+                    consumedRA.setAssigneeIdentifier(baseAU.getIdentifier());
+                    entityManager.merge(consumedRA);
+                    solrIndexService.indexPermissionsForOneDvObject(consumedRA.getDefinitionPoint());
+                    indexService.indexDvObject(consumedRA.getDefinitionPoint());
+                } // no else here because the any willDelete == true will happen in the named query below.
+            } else {
+                throw new IllegalArgumentException("Original userIdentifier provided does not seem to be an AuthenticatedUser");
+            }
+        }
+
+        //Delete not merged role assignments for consumedIdentifier, e.g. duplicates
+        entityManager.createNamedQuery("RoleAssignment.deleteAllByAssigneeIdentifier", RoleAssignment.class)
+                .setParameter("assigneeIdentifier", consumedAU.getIdentifier())
+                .executeUpdate();
+    }
+
+    private void updateDatasetVersionUser(AuthenticatedUser consumedAU, AuthenticatedUser baseAU) {
+        for (DatasetVersionUser user : datasetDao.getDatasetVersionUsersByAuthenticatedUser(consumedAU)) {
+            user.setAuthenticatedUser(baseAU);
+            entityManager.merge(user);
+        }
+    }
+
+    private void updateDatasetLocks(AuthenticatedUser consumedAU, AuthenticatedUser baseAU) {
+        for (DatasetLock lock : datasetDao.getDatasetLocksByUser(consumedAU)) {
+            lock.setUser(baseAU);
```

Comment:
I would check if all of these loops are required, since in our case usually collections have Cascade option. So if you set locks in Authenticated user it should propagate in DatasetLock. Anyway if you are updating in current way you also have to update AuthenticatedUser side afaik.

Reply:
I understand that some of these calls are redundant, but I'd still leave it as is. 
This is important administrative task and as such I want to be certain that all entities are handled explicitly and not to depend on whether or not `CASCADE` was changed somewhere so it would no longer work.

What do you mean by AuthenticatedUser side here? We take care of authenticated users.

Reply:
Well if you wanna be certain then you have to have Integration test to check it, how are you sure now that it works otherwise? You are using some kind of Framework so you are bound to trust it.

Well AuthenticatedUser side is not taken care of, if you query user then he won't have DatasetLock reference for example, just check in test, you can query datasetlock with user reference but user doesnt have DatasetLock reference so it doesn't work properly.

Reply:
I don't want to test database integrity. I only and explicitly check if all db entries regarding user are being taken care of. Nothing more.

Reply:
But db entries regarding user are not taken care of, the user doesn't have references!

Reply:
Point me a single entry that won't update properly then because i don't see it.

Reply:
AuthenticatedUser - datasetLocks

Reply:
You are mistaking db entries with what you see in Java. There is no entry `AuthenticatedUser - datasetLocks`. 
For `DatasetLock` we update user such as lock is now being owned by `base` user, then we remove `consumed` user.

Reply:
Yes, I'm not talking exactly from db perspective but entities, If you query AuthenticatedUser in your IT he won't have datasetlocks even if datasetlock has the user.

Reply:
I understand that from db perspective it's ok, but for some reason it doesn't work in IT.

Reply:
In IT I don't care much about what entities include. I only test db state before and after merge.
If its valid, then entities must reflect db state properly after.